### PR TITLE
bootctl: 'graceful' should ignore EFI variable failures

### LINKED ISSUE #26773: 'bootctl update --graceful' does not ignore EFI variable errors as documented
### systemd version the issue has been seen with

systemd_252.6-1_amd64.deb

### Used distribution

Debian Bookworm

### Linux kernel version used

6.1.0-6-amd64

### CPU architectures issue was seen on

arm

### Component

bootctl

### Expected behaviour you didn't see

`bootctl update --graceful` should complete without errors on a system which does not support EFI variables (U-boot in this case).

### Unexpected behaviour you saw

`bootctl update --graceful` reported a failure to access EFI variables and exited with a non-zero status.

### Steps to reproduce the problem

Run `bootctl update --graceful` on a system which does not support EFI variables and has systemd-boot already installed (using `bootctl install --no-variables`).

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
root@core23-a:~# bootctl update --graceful
Skipping "/efi/EFI/systemd/systemd-bootaa64.efi", since same boot loader version in place already.
Skipping "/efi/EFI/BOOT/BOOTAA64.EFI", since same boot loader version in place already.
Failed to create EFI Boot variable entry: No such file or directory
```

### PR DESCRIPTION
**Pull request by kpfleming**
Suppress errors when creating/writing EFI variables during 'bootctl update' if '--graceful' mode is active (as the documentation indicates).

Closes #26773.

### CODE REVIEW

**Comment by bluca** [comment]
It should also return success, not just suppress the log message

**Comment by kpfleming** [comment]
First push was broken as it would have also suppressed errors during `install`... Second push has the correct code.

**Comment by kpfleming** [comment]
@bluca Hmm... I copied the pattern from `acquire_esp` in bootctl.c, which is assuming that `log_full_errno` will return 0. That may not be a safe assumption, in which case `acquire_esp` is also broken.

**Comment by bluca** [comment]
Can you check what the exit code is in that case?

**Comment by kpfleming** [comment]
Tested and you were correct; status code was non-zero even in graceful mode. This last push corrects that.

**Comment by poettering** [comment]
i don't get this. what is an uefi system without efi vars? the concept is not precisely optional in the uefi spec, is it?

sounds fine to me to handle gracefully if we can't write efi vars, but a system where efi vars don't exist at all seems super broken. it's also fine if persistent efi vars don't exist, but ultimately any firmware can implement them just in RAM if they want, and if they don't that's super broken.

This sounds like something to fix in your firmware? what am i missing? why make this frankenuefi our problem?

**Comment by bluca** [comment]
Firmwares do wonky things in the real world - `--graceful` is documented as ignoring failures due to variables, but that's plainly not happening, so I think it makes sense to fix it to make it do what it says on the tin

**Comment by bluca** [comment]
Or alternatively think about the use case where there's no variables because it's running in a container to update an image, or such cases

**Comment by poettering** [comment]
> Firmwares do wonky things in the real world - `--graceful` is documented as ignoring failures due to variables, but that's plainly not happening, so I think it makes sense to fix it to make it do what it says on the tin

It is not? what precisely *is* happening? the report never mention what precisely is happening.

And the docs clearly say "…when EFI variables cannot be written…". i.e. nothing about read, only about write.

**Comment by poettering** [comment]
> Or alternatively think about the use case where there's no variables because it's running in a container to update an image, or such cases

Well, detect the container case automatically already.

**Comment by bluca** [comment]
> > Firmwares do wonky things in the real world - `--graceful` is documented as ignoring failures due to variables, but that's plainly not happening, so I think it makes sense to fix it to make it do what it says on the tin
> 
> It is not? what precisely _is_ happening? the report never mention what precisely is happening.

The error report has the example:

```
root@core23-a:~# bootctl update --graceful
Skipping "/efi/EFI/systemd/systemd-bootaa64.efi", since same boot loader version in place already.
Skipping "/efi/EFI/BOOT/BOOTAA64.EFI", since same boot loader version in place already.
Failed to create EFI Boot variable entry: No such file or directory
```

> And the docs clearly say "…when EFI variables cannot be written…". i.e. nothing about read, only about write.

They are read as part of the process of updating them, given it's the `update` verb

File: src/boot/bootctl-install.c [hunk 1]
```diff
@@ -671,18 +672,28 @@ static int install_variables(
                 return log_error_errno(r, "Cannot access \"%s/%s\": %m", esp_path, path);
 
         r = find_slot(uuid, path, &slot);
-        if (r < 0)
+        if (r < 0) {
+                if (graceful)
+                        return log_full_errno(arg_quiet ? LOG_DEBUG : LOG_INFO, r,
+                                              "Failed to access EFI variables, skipping.");
```
Comment (keszybz) [inline src/boot/bootctl-install.c]: ": %m"

File: src/boot/bootctl-install.c [hunk 2]
```diff
@@ -671,18 +672,28 @@ static int install_variables(
...
 
         if (first || r == 0) {
                 r = efi_add_boot_option(slot, pick_efi_boot_option_description(),
                                         part, pstart, psize,
                                         uuid, path);
-                if (r < 0)
+                if (r < 0) {
+                        if (graceful)
+                                return log_full_errno(arg_quiet ? LOG_DEBUG : LOG_INFO, r,
+                                                      "Failed to create EFI Boot variable entry, skipping.");
```
Comment (keszybz) [inline src/boot/bootctl-install.c]: ": %m" here too.

**Review by keszybz — CHANGES_REQUESTED** [review]
This looks reasonable, but needs a little cleanup.

With the patch, both the caller suppresses the return value (if `graceful`), and the callee also does this internally. If you're changing the callee to suppress errors, then also make it return 0 so that the caller does not need to be changed.

**Comment by kpfleming** [comment]
I can certainly make `install_variables` suppress the errno return instead of making the caller do it; I chose this pattern as it was what was already being done in other places in the code (in particular calls to `acquire_esp`). This allows the caller to decide whether the particular errno returned is covered by `graceful` completely or not.

File: src/boot/bootctl-install.c (hunk 1 above)
Comment (kpfleming) [reply src/boot/bootctl-install.c]: Done.

File: src/boot/bootctl-install.c (hunk 2 above)
Comment (kpfleming) [reply src/boot/bootctl-install.c]: Done.

**Comment by keszybz** [comment]
> I can certainly make `install_variables` suppress the errno return instead of making the caller do it; I chose this pattern as it was what was already being done in other places in the code (in particular calls to `acquire_esp`). This allows the caller to decide whether the particular errno returned is covered by `graceful` completely or not.

Right now we can have a situation where `install_variables` calls `log_error`, but the caller than then suppresses the error. This is too complicated and error-prone. Please rework `install_variables` to return 0 if `graceful` and the error should be suppressed.

**Comment by kpfleming** [comment]
Done. I also ensured that the logged messages are the same in both graceful and non-graceful modes when the EFI variables cannot be accessed at all.

File: src/boot/bootctl-install.c [hunk 3]
```diff
@@ -671,18 +672,34 @@ static int install_variables(
...
+                                       r == -ENOENT ?
+                                       "Failed to access EFI variables, skipping. Is the \"efivarfs\" filesystem mounted?" :
+                                       "Failed to determine current boot order, skipping: %m");
+                        return 0;
+                }
+
                 return log_error_errno(r,
                                        r == -ENOENT ?
                                        "Failed to access EFI variables. Is the \"efivarfs\" filesystem mounted?" :
                                        "Failed to determine current boot order: %m");
```
Comment (keszybz) [inline src/boot/bootctl-install.c]: OK. This should work, but now both the graceful and non-graceful path are almost the same. Please reformat:
```C
int level = graceful ? arg_quiet ? LOG_DEBUG : LOG_INFO : LOG_ERR;
const char *skip = graceful ? ", skipping" : "";
log_full_errno(level, r,
                          r == -ENOENT ? … : …,
                          skip);
return graceful ? 0 : r;
```

And similarly for the case below.

Comment (kpfleming) [reply src/boot/bootctl-install.c]: Thanks for the gentle nudge... my C skills are rusty and I hadn't considered refactoring the code that way. It's much clearer now.

**Pull request merged**